src/connection.sml

```diff
@@ -1,68 +1,194 @@
+        , serializedQuery
+        , Word8VectorSlice.full (convertString "\000")
+        ])
+      val str = PolyML.makestring message
+    in
+      print str;
+      Socket.sendVec (socket, message);
+      OK
+    end
+
+  fun parser (socket: Socket.active INetSock.stream_sock) =
+    let
+      val messageType = (* Byte.bytesToString ( *) Socket.recvVec (socket, 1)
+      val length = (* bytesToInt( *) Socket.recvVec (socket, 4)
+      val mP =
+        "Type: " ^ PolyML.makestring messageType ^ " -> "
+        ^ (Byte.bytesToString messageType)
+      val lP =
+        "Length: " ^ PolyML.makestring length ^ " -> "
+        ^ (Int.toString (bytesToInt length))
+    in
+      print "\n\n";
+      print mP;
+      print "\n\n";
+      print lP;
+      print "\n\n";
+      case (Byte.bytesToString messageType) of
+        "R" =>
+          ( print "Authenticating\n"
+          ; password (Socket.recvVec (socket, (bytesToInt length) - 4))
+          ; Posix.Process.sleep (Time.fromSeconds 1)
+          ; parser socket
+          )
+      | "Z" =>
+          ( print "ReadyForQuery\n"
+          ; Posix.Process.sleep (Time.fromSeconds 1)
+          ; print
+              ("Transaction status: "
+               ^ (Byte.bytesToString (Socket.recvVec (socket, 1))) ^ "\n")
+          )
+      | "E" =>
+          ( print
+              ("ERROR: "
+               ^
+               (Byte.bytesToString (Socket.recvVec
+                  (socket, (bytesToInt length) - 4))) ^ "\n")
+          ; Posix.Process.sleep (Time.fromSeconds 1)
+          ; parser socket
+          )
+      | "S" =>
+          ( print "Syncing...\n"
+          ; Socket.recvVec (socket, (bytesToInt length) - 4)
+          ; (* Posix.Process.sleep (Time.fromSeconds 1); *) parser socket
+          )
+      | "K" =>
+          ( print "BackendKeyData\n"
+          ; Socket.recvVec (socket, (bytesToInt length) - 4)
+          ; parser socket
+          )
+      | "T" =>
+          (let
+             val numberOfHeaders = bytesToInt16 (Socket.recvVec (socket, 2))
+             val contents = ref (Word8VectorSlice.full (Socket.recvVec
+               (socket, (bytesToInt length) - 4 - 2)))
+             val fieldName = ref ""
+           in
+             print ("Receiving " ^ Int.toString numberOfHeaders ^ " headers\n");
+             if numberOfHeaders <> 0 then
+               (case
+                  Word8VectorSlice.findi (fn (_, byte) => byte = 0wx0)
+                    (!contents)
+                of
+                  SOME (index, _) =>
+                    ( fieldName
+                      :=
+                      Byte.bytesToString (Word8VectorSlice.concat
+                        [Word8VectorSlice.subslice ((!contents), 0, SOME index)])
+                    ; print ("FIELD: " ^ (!fieldName) ^ "\n")
+                    )
+                | NONE => ())
+             else
+               print "Nothing :(\n"
+           end
+           handle SysErr => print "SYS\n")
```

Comment:
should'd you print the exn here?

Reply:
The exception is simply SysErr in that case, so I would have to do a PolyML.makestring just to print "SysErr". I am better off handling one by one and writing an explanation